arwen/elrondapi/elrondei.go

```diff
@@ -627,3 +656,280 @@ func int64finish(context unsafe.Pointer, value int64) {
+
+	currOffset := dataOffset
+	data := make([][]byte, numArguments)
+	for i := int32(0); i < numArguments; i++ {
+		currArgLenData := argumentsLengthData[i*4 : i*4+4]
+		actualLen := dataToInt32(currArgLenData)
+
+		data[i] = arwen.LoadBytes(instCtx.Memory(), currOffset, actualLen)
+		currOffset += actualLen
+	}
+
+	function := arwen.LoadBytes(instCtx.Memory(), functionOffset, functionLength)
+
+	return string(function), data, currOffset - dataOffset
+}
+
+//export delegateExecution
+func delegateExecution(
+	context unsafe.Pointer,
+	gasLimit int64,
+	addressOffset int32,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) int32 {
+	instCtx := wasmer.IntoInstanceContext(context)
+	erdContext := arwen.GetErdContext(instCtx.Data())
+
+	address := arwen.LoadBytes(instCtx.Memory(), addressOffset, arwen.HashLen)
+	function, data, actualLen := getArgumentsFromMemory(context, functionOffset, functionLength, numArguments, argumentsLengthOffset, dataOffset)
+
+	value := erdContext.GetVMInput().CallValue
+	sender := erdContext.GetVMInput().CallerAddr
+
+	gasToUse := erdContext.GasSchedule().ElrondAPICost.DelegateExecution
+	gasToUse += erdContext.GasSchedule().BaseOperationCost.StorePerByte * uint64(actualLen)
+	erdContext.UseGas(gasToUse)
+
+	if erdContext.GasLeft() < uint64(gasLimit) {
+		return 1
+	}
+
+	erdContext.Transfer(address, sender, value, nil)
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  sender,
+			Arguments:   data,
+			CallValue:   value,
+			GasPrice:    0,
+			GasProvided: uint64(gasLimit),
+		},
+		RecipientAddr: address,
+		Function:      function,
+	}
+	err := erdContext.ExecuteOnSameContext(contractCallInput)
+	if err != nil {
+		return 1
+	}
+
+	return 0
+}
+
+func dataToInt32(data []byte) int32 {
+	actualLen := int32(0)
+	for i := len(data) - 1; i >= 0; i-- {
+		actualLen = (actualLen << 8) + int32(data[i])
+	}
+
+	return actualLen
+}
+
+//export executeReadOnly
+func executeReadOnly(
+	context unsafe.Pointer,
+	gasLimit int64,
+	addressOffset int32,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) int32 {
+	instCtx := wasmer.IntoInstanceContext(context)
+	erdContext := arwen.GetErdContext(instCtx.Data())
+
+	address := arwen.LoadBytes(instCtx.Memory(), addressOffset, arwen.HashLen)
+	function, data, actualLen := getArgumentsFromMemory(context, functionOffset, functionLength, numArguments, argumentsLengthOffset, dataOffset)
+
+	value := erdContext.GetVMInput().CallValue
+	sender := erdContext.GetVMInput().CallerAddr
+
+	gasToUse := erdContext.GasSchedule().ElrondAPICost.ExecuteReadOnly
+	gasToUse += erdContext.GasSchedule().BaseOperationCost.StorePerByte * uint64(actualLen)
+	erdContext.UseGas(gasToUse)
+
+	if erdContext.GasLeft() < uint64(gasLimit) {
+		return 1
+	}
+
```

Comment:
Probably need `gasLimit := erdContext.GasLeft()` here.

Reply:
no, gasLimit is input for function